File: backend/app/schemas/pipeline.py
import uuid

from pydantic import BaseModel, Field


class PipelineStageCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    position: int = 0
    is_terminal: bool = False


class PipelineStageResponse(BaseModel):
    id: uuid.UUID
    name: str
    position: int
    is_terminal: bool

    model_config = {"from_attributes": True}


class MoveClientStageRequest(BaseModel):
    client_id: uuid.UUID
    stage_id: uuid.UUID
    assigned_user_id: uuid.UUID | None = None


class KanbanClientItem(BaseModel):
    id: uuid.UUID
    name: str
    city: str | None = None
    country: str | None = None
    rating: float | None = None
    category: str | None = None
    priority: str = "medium"
    nextFollowUp: str | None = None
    assignedRep: str | None = None
    phone: str | None = None
    email: str | None = None

    model_config = {"from_attributes": True}


class KanbanColumnResponse(BaseModel):
    stage: PipelineStageResponse
    clients: list[KanbanClientItem]

File: frontend/lib/api/pipeline.ts
import { apiFetch } from "./client";

export interface PipelineStageResponse {
  id: string;
  org_id: string;
  name: string;
  slug: string;
  position: number;
  color: string | null;
}

export interface PipelineStageCreateRequest {
  name: string;
  slug: string;
  position: number;
  color?: string;
}

export interface KanbanClientItem {
  id: string;
  name: string;
  city: string | null;
  country: string | null;
  rating: number | null;
  category?: string | null;
  priority?: "low" | "medium" | "high" | string;
  nextFollowUp?: string | null;
  assignedRep?: string | null;
  phone?: string | null;
  email?: string | null;
}

export interface KanbanColumnResponse {
  stage: PipelineStageResponse;
  clients: KanbanClientItem[];
}

export interface MoveClientStageRequest {
  client_id: string;
  stage_id: string;
}

export async function listStages(): Promise<PipelineStageResponse[]> {
  return apiFetch<PipelineStageResponse[]>("/api/v1/pipeline/stages");
}

export async function createStage(
  payload: PipelineStageCreateRequest
): Promise<PipelineStageResponse> {
  return apiFetch<PipelineStageResponse>("/api/v1/pipeline/stages", {
    method: "POST",
    body: payload,
  });
}

export async function getBoard(): Promise<KanbanColumnResponse[]> {
  return apiFetch<KanbanColumnResponse[]>("/api/v1/pipeline/board");
}

export async function moveClient(
  payload: MoveClientStageRequest
): Promise<void> {
  return apiFetch<void>("/api/v1/pipeline/move", {
    method: "POST",
    body: payload,
  });
}
